File: skills/plotting/python/pavlab_scatter.py
"""pavlab_scatter — opinionated lab-style scatter plot.

Defaults: black filled circles, no background, no gridlines, large axis
labels. Automatically adapts rendering to N:
  N ≤ 500         full-opacity circles (s=20)
  500 < N ≤ 10 000 semi-transparent circles (alpha scales 0.60 → 0.15)
  N > 10 000       hexbin density (black-body palette)

Usage
-----
    from pavlab_scatter import pavlab_scatter

    # basic
    pavlab_scatter(x, y)

    # log-scaled axes — label gets "(log₂)" suffix automatically
    pavlab_scatter(x, y, xlabel="TPM rep1", ylabel="TPM rep2",
                   log_x="log2", log_y="log2")

    # categorical colour → lab accent palette + legend
    pavlab_scatter(x, y, color=cell_type_labels, xlabel="PC1", ylabel="PC2")

    # numeric colour → viridis colorbar
    pavlab_scatter(x, y, color=scores, color_label="score")

    # save
    pavlab_scatter(x, y, filename="fig3a.pdf", figsize=(4, 4))

    # embed in a multi-panel figure
    fig, axes = plt.subplots(1, 2, figsize=(10, 4))
    pavlab_scatter(x1, y1, ax=axes[0])
    pavlab_scatter(x2, y2, ax=axes[1])
"""

from __future__ import annotations

import warnings
from typing import Any, Union

import matplotlib.pyplot as plt
import numpy as np

try:
    from .palettes import black_body_palette
except ImportError:
    from palettes import black_body_palette  # script / sys.path style

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_BLACK = "#000000"

# Lab accent palette (Tailwind-style, matches global lab convention)
_ACCENT_COLORS = [
    "#2563eb",  # blue-600   – primary
    "#10b981",  # emerald-500
    "#f59e0b",  # amber-500
    "#ef4444",  # red-500
    "#8b5cf6",  # violet-500
    "#ec4899",  # pink-500
    "#14b8a6",  # teal-500
    "#f97316",  # orange-500
]

_N_ALPHA: int = 500        # above this N: add transparency
_N_DENSITY: int = 10_000   # above this N: switch to hexbin

# (axis-label pt, tick-label pt)
_LABEL_SIZES: dict[str, tuple[float, float]] = {
    "small": (10.0,  8.0),
    "med":   (14.0, 11.0),
    "large": (18.0, 14.0),
}

LogScale = Union[bool, str]   # False | True | "log2" | "log10"


# ---------------------------------------------------------------------------
# Private helpers
# ---------------------------------------------------------------------------

def _pick_log_kind(val: LogScale, arr: np.ndarray, axis: str) -> str | None:
    """Resolve a LogScale value to 'log2', 'log10', or None (no transform)."""
    if val is False or val is None:
        return None
    if val is True:
        pos = arr[np.isfinite(arr) & (arr > 0)]
        if pos.size == 0:
            warnings.warn(
                f"pavlab_scatter: {axis}-axis has no positive values; "
                "log transform skipped.",
                stacklevel=3,
            )
            return None
        # Heuristic: large counts → log10; biological/moderate range → log2.
        return "log10" if float(pos.max()) > 1_000 else "log2"
    if val in ("log2", "log10"):
        return str(val)
    raise ValueError(
        f"log_{axis} must be False, True, 'log2', or 'log10'; got {val!r}"
    )


def _apply_log(arr: np.ndarray, kind: str, pseudocount: float) -> np.ndarray:
    """Apply log2 or log10 after adding pseudocount."""
    shifted = arr + pseudocount
    with np.errstate(divide="ignore", invalid="ignore"):
        return np.log2(shifted) if kind == "log2" else np.log10(shifted)


def _axis_label(base: str | None, log_kind: str | None) -> str:
    """'FPKM' + 'log2' → 'FPKM (log₂)'; None + 'log2' → '(log₂)'."""
    suffix_map = {"log2": "(log₂)", "log10": "(log₁₀)"}
    parts = [p for p in [base, suffix_map.get(log_kind or "")] if p]
    return " ".join(parts)


def _font_sizes(label_size: str | float) -> tuple[float, float]:
    """Return (axis_label_pt, tick_label_pt)."""
    if isinstance(label_size, str):
        try:
            return _LABEL_SIZES[label_size]
        except KeyError:
            raise ValueError(
                f"label_size must be 'small', 'med', 'large', or a number; "
                f"got {label_size!r}"
            )
    v = float(label_size)
    return v, round(v * 0.78, 1)


def _maybe_warn_log(arr: np.ndarray, axis: str) -> None:
    """Suggest log scale when data spans more than two orders of magnitude."""
    pos = arr[np.isfinite(arr) & (arr > 0)]
    if pos.size < 2:
        return
    ratio = float(pos.max()) / float(pos.min())
    if ratio > 100:
        warnings.warn(
            f"pavlab_scatter: {axis}-axis spans {ratio:.0f}× — "
            f"consider log_{axis}='log2' (biological) or 'log10' (counts).",
            stacklevel=3,
        )


def _resolve_color(color, n: int):
    """Parse the color argument.

    Returns (c_arg, cmap, norm, legend_handles, is_single_color).
    When is_single_color=True use `color=` kwarg to ax.scatter (not `c=`).
    """
    import matplotlib.patches as mpatches
    from matplotlib.colors import Normalize

    if color is None:
        return _BLACK, None, None, None, True  # uniform black → color=

    arr = np.asarray(color)

    # Categorical (string / object dtype) → lab accent palette + legend
    if arr.dtype.kind in ("U", "S", "O"):
        categories = list(dict.fromkeys(arr.tolist()))
        cat_color = {
            cat: _ACCENT_COLORS[i % len(_ACCENT_COLORS)]
            for i, cat in enumerate(categories)
        }
        c_list = [cat_color[v] for v in arr.tolist()]
        handles = [
            mpatches.Patch(color=cat_color[cat], label=str(cat))
            for cat in categories
        ]
        return c_list, None, None, handles, False

    # Numeric → viridis colormap
    norm = Normalize(vmin=float(np.nanmin(arr)), vmax=float(np.nanmax(arr)))
    return arr, "viridis", norm, None, False


# ---------------------------------------------------------------------------
# Main function
# ---------------------------------------------------------------------------

def pavlab_scatter(
    x,
    y,
    color=None,
    color_label: str | None = None,
    label_size: str | float = "med",
    xlabel: str | None = None,
    ylabel: str | None = None,
    title: str | None = None,
    log_x: LogScale = False,
    log_y: LogScale = False,
    pseudocount: float = 1.0,
    origin_zero: bool = True,
    xlim: tuple | None = None,
    ylim: tuple | None = None,
    point_size: float | None = None,
    filename=None,
    figsize: tuple | None = None,
    ax=None,
    **kwargs: Any,
):
    """Pavlab-style scatter plot.

    Parameters
    ----------
    x, y : array-like
        Data coordinates. NaN / inf pairs are silently dropped.
    color : None | str | array-like, optional
        None → all black. A single matplotlib color string → uniform color.
        An array of strings → categorical colors (lab accent palette, legend
        added automatically). An array of numbers → viridis colormap +
        colorbar. In density mode (N > 10 000) color is ignored.
    color_label : str, optional
        Colorbar label for continuous color. Ignored for categorical color.
    label_size : "small" | "med" | "large" | float
        Axis-label font size preset. "large" (18 pt labels / 14 pt ticks)
        survives projector display and journal reduction to a 3-inch column.
    xlabel, ylabel : str, optional
        Axis labels. Log-scale suffix ("(log₂)" / "(log₁₀)") is appended
        automatically when a transform is applied.
    title : str, optional
        Figure title. Left-aligned, normal weight.
    log_x, log_y : False | True | "log2" | "log10"
        Log-transform the axis. True auto-selects: log10 when max > 1 000
        (counts), log2 otherwise (biological / expression). pseudocount is
        added before transforming (handles zeros). ln is never used.
        When not transforming, a warning is issued if data spans > 100×.
    pseudocount : float
        Added to values before log transform. Default 1.
    origin_zero : bool
        If True (default) and all (post-transform) data ≥ 0, pin the axis
        lower bound to 0. Explicit xlim= / ylim= take precedence.
    xlim, ylim : (lo, hi) | None
        Explicit axis limits. Take precedence over origin_zero.
    point_size : float | None
        Override the automatic marker size (matplotlib `s=`). Default: 20
        for N ≤ 500, 10 for N ≤ 10 000.
    filename : str | Path | None
        Save to file; format inferred from extension (.pdf, .png, .svg …).
        Ignored when ax= is supplied.
    figsize : (w, h) | None
        Figure size in inches. Default (5.5, 5.0). Ignored when ax= is
        supplied.
    ax : matplotlib Axes | None
        Draw into an existing Axes. figsize and filename are ignored.
    **kwargs
        Forwarded to ax.scatter (or ax.hexbin in density mode).

    Returns
    -------
    matplotlib Axes
    """
    # ---- coerce to 1-D float arrays -------------------------------------
    x_arr = np.asarray(x, dtype=float).ravel()
    y_arr = np.asarray(y, dtype=float).ravel()
    if len(x_arr) != len(y_arr):
        raise ValueError(
            f"x and y must have equal length ({len(x_arr)} vs {len(y_arr)})"
        )

    # Align per-point color with x/y, then drop non-finite pairs together.
    color_arr: Any = None
    c_tmp = np.asarray(color) if color is not None else None
    if c_tmp is not None and c_tmp.ndim == 1 and len(c_tmp) == len(x_arr):
        mask = np.isfinite(x_arr) & np.isfinite(y_arr)
        color_arr = c_tmp[mask]
        x_arr, y_arr = x_arr[mask], y_arr[mask]
    else:
        mask = np.isfinite(x_arr) & np.isfinite(y_arr)
        x_arr, y_arr = x_arr[mask], y_arr[mask]
        color_arr = color  # scalar or None — pass through unchanged

    n = len(x_arr)
    if n == 0:
        warnings.warn("pavlab_scatter: no finite data points to plot.", stacklevel=2)

    # ---- suggest log scale when not transforming ------------------------
    if log_x is False:
        _maybe_warn_log(x_arr, "x")
    if log_y is False:
        _maybe_warn_log(y_arr, "y")

    # ---- apply log transforms ------------------------------------------
    x_kind = _pick_log_kind(log_x, x_arr, "x")
    if x_kind:
        neg = int(np.sum(np.isfinite(x_arr) & (x_arr + pseudocount <= 0)))
        if neg:
            warnings.warn(
                f"pavlab_scatter: {neg} x value(s) will produce -inf after "
                f"log transform (pseudocount={pseudocount} insufficient).",
                stacklevel=2,
            )
        x_arr = _apply_log(x_arr, x_kind, pseudocount)

    y_kind = _pick_log_kind(log_y, y_arr, "y")
    if y_kind:
        neg = int(np.sum(np.isfinite(y_arr) & (y_arr + pseudocount <= 0)))
        if neg:
            warnings.warn(
                f"pavlab_scatter: {neg} y value(s) will produce -inf after "
                f"log transform (pseudocount={pseudocount} insufficient).",
                stacklevel=2,
            )
        y_arr = _apply_log(y_arr, y_kind, pseudocount)

    # ---- rendering mode ------------------------------------------------
    density_mode = n > _N_DENSITY
    if density_mode and color is not None:
        warnings.warn(
            "pavlab_scatter: N > 10 000; using hexbin density mode — "
            "'color' ignored.",
            stacklevel=2,
        )
        color_arr = None

    # ---- colour resolution ---------------------------------------------
    c_arg = cmap_arg = norm_arg = legend_handles = None
    is_single = True
    if not density_mode:
        c_arg, cmap_arg, norm_arg, legend_handles, is_single = \
            _resolve_color(color_arr, n)

    # ---- marker size + alpha -------------------------------------------
    ps = float(point_size) if point_size is not None else (
        20.0 if n <= _N_ALPHA else 10.0
    )
    if n <= _N_ALPHA:
        alpha = 1.0
    else:
        frac = min(1.0, (n - _N_ALPHA) / max(1, _N_DENSITY - _N_ALPHA))
        alpha = max(0.15, 0.6 - frac * 0.45)

    # ---- figure setup --------------------------------------------------
    own_fig = ax is None
    if own_fig:
        fig, ax = plt.subplots(figsize=figsize or (5.5, 5.0))
    else:
        fig = ax.figure

    ax.set_facecolor("white")
    ax.grid(False)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)

    # ---- draw ----------------------------------------------------------
    label_pt, tick_pt = _font_sizes(label_size)

    if density_mode:
        hb_kw: dict[str, Any] = {
            "cmap": black_body_palette(256),
            "mincnt": 1,
            "gridsize": 40,
        }
        hb_kw.update(kwargs)
        hb = ax.hexbin(x_arr, y_arr, **hb_kw)
        cb = fig.colorbar(hb, ax=ax)
        cb.set_label("count", fontsize=tick_pt)

    else:
        sc_kw: dict[str, Any] = {
            "s": ps,
            "alpha": alpha,
            "linewidths": 0,
            "marker": "o",
        }
        if is_single:
            sc_kw["color"] = c_arg   # uniform color: use color= not c=
        else:
            sc_kw["c"] = c_arg
            if cmap_arg is not None:
                sc_kw["cmap"] = cmap_arg
            if norm_arg is not None:
                sc_kw["norm"] = norm_arg
        sc_kw.update(kwargs)
        sc = ax.scatter(x_arr, y_arr, **sc_kw)

        if cmap_arg is not None:
            cb = fig.colorbar(sc, ax=ax)
            if color_label:
                cb.set_label(color_label, fontsize=tick_pt)

        if legend_handles:
            ax.legend(handles=legend_handles, frameon=False, fontsize=tick_pt)

    # ---- axis limits ---------------------------------------------------
    if xlim is not None:
        ax.set_xlim(xlim)
    else:
        fx = x_arr[np.isfinite(x_arr)]
        if origin_zero and fx.size > 0 and float(fx.min()) >= 0:
            ax.set_xlim(left=0)

    if ylim is not None:
        ax.set_ylim(ylim)
    else:
        fy = y_arr[np.isfinite(y_arr)]
        if origin_zero and fy.size > 0 and float(fy.min()) >= 0:
            ax.set_ylim(bottom=0)

    # ---- labels --------------------------------------------------------
    xl = _axis_label(xlabel, x_kind)
    yl = _axis_label(ylabel, y_kind)
    if xl:
        ax.set_xlabel(xl, fontsize=label_pt)
    if yl:
        ax.set_ylabel(yl, fontsize=label_pt)
    if title:
        ax.set_title(title, fontsize=label_pt, loc="left", fontweight="normal")

    ax.tick_params(axis="both", labelsize=tick_pt)

    # ---- output --------------------------------------------------------
    if own_fig:
        fig.tight_layout()
        if filename is not None:
            fig.savefig(filename, dpi=200)
            plt.close(fig)

    return ax

File: skills/plotting/R/pavlab_scatter.R
# pavlab_scatter: opinionated lab-style scatter plot (ggplot2 backend).
# Mirrors Python pavlab_scatter defaults exactly:
#   - Black filled circles, no background, no gridlines, top+right off
#   - N-adaptive: solid circles ≤500, alpha-scaled 501–10 000, hexbin >10 000
#   - log_x/log_y: FALSE/TRUE/"log2"/"log10"; warns when data spans >100x; never ln
#   - origin_zero: pins non-negative axis lower bound to 0
#   - color: NULL=black; single string=uniform; char/factor vec=categorical+legend;
#            numeric vec=viridis colorbar
#
# Usage:
#   source("R/palettes.R")
#   source("R/pavlab_scatter.R")
#   pavlab_scatter(x, y)
#   pavlab_scatter(x, y, log_x="log2", log_y="log2", xlabel="TPM rep1", ylabel="TPM rep2")
#   pavlab_scatter(x, y, color=cell_types, xlabel="PC1", ylabel="PC2", origin_zero=FALSE)
#   p <- pavlab_scatter(x, y, filename="fig3a.pdf")
#   p + ggplot2::geom_abline(slope=1, linetype="dashed")   # extend with ggplot2 layers
#
# Dependencies: ggplot2 (>= 3.4.0); hexbin required only for N > 10 000.

# ---- Constants ---------------------------------------------------------------

.N_ALPHA   <- 500L
.N_DENSITY <- 10000L

# (axis_label_pt, tick_pt)
.LABEL_SIZES <- list(
  small = c(10.0,  8.0),
  med   = c(14.0, 11.0),
  large = c(18.0, 14.0)
)

.ACCENT_COLORS <- c(
  "#2563eb",  # blue-600   – primary
  "#10b981",  # emerald-500
  "#f59e0b",  # amber-500
  "#ef4444",  # red-500
  "#8b5cf6",  # violet-500
  "#ec4899",  # pink-500
  "#14b8a6",  # teal-500
  "#f97316"   # orange-500
)

# ---- Private helpers ---------------------------------------------------------

.pick_log_kind <- function(val, arr, axis) {
  if (identical(val, FALSE) || is.null(val)) return(NULL)
  if (identical(val, TRUE)) {
    pos <- arr[is.finite(arr) & arr > 0]
    if (length(pos) == 0) {
      warning(sprintf(
        "pavlab_scatter: %s-axis has no positive values; log transform skipped.",
        axis
      ), call. = FALSE)
      return(NULL)
    }
    return(if (max(pos) > 1000) "log10" else "log2")
  }
  if (is.character(val) && val %in% c("log2", "log10")) return(val)
  stop(sprintf(
    "log_%s must be FALSE, TRUE, 'log2', or 'log10'; got '%s'", axis, val
  ), call. = FALSE)
}

.apply_log <- function(arr, kind, pseudocount) {
  shifted <- arr + pseudocount
  if (kind == "log2") log2(shifted) else log10(shifted)
}

.axis_label <- function(base, log_kind) {
  # Unicode subscripts: log₂ = log₂, log₁₀ = log₁₀
  suffix_map <- c(log2 = "(log₂)", log10 = "(log₁₀)")
  suffix <- if (!is.null(log_kind) && log_kind %in% names(suffix_map))
               suffix_map[[log_kind]] else NULL
  parts <- character(0)
  if (!is.null(base)    && nchar(base)   > 0) parts <- c(parts, base)
  if (!is.null(suffix)  && nchar(suffix) > 0) parts <- c(parts, suffix)
  paste(parts, collapse = " ")
}

.font_sizes <- function(label_size) {
  if (is.character(label_size)) {
    if (!label_size %in% names(.LABEL_SIZES))
      stop(sprintf(
        "label_size must be 'small', 'med', 'large', or a number; got '%s'",
        label_size
      ), call. = FALSE)
    return(.LABEL_SIZES[[label_size]])
  }
  v <- as.numeric(label_size)
  c(v, round(v * 0.78, 1))
}

.maybe_warn_log <- function(arr, axis) {
  pos <- arr[is.finite(arr) & arr > 0]
  if (length(pos) < 2) return(invisible(NULL))
  ratio <- max(pos) / min(pos)
  if (ratio > 100) {
    warning(sprintf(
      "pavlab_scatter: %s-axis spans %.0fx — consider log_%s='log2' (biological) or 'log10' (counts).",
      axis, ratio, axis
    ), call. = FALSE)
  }
  invisible(NULL)
}

.color_mode <- function(color, n) {
  if (is.null(color))                                        return("uniform_black")
  if (is.character(color) && length(color) == 1)             return("uniform_string")
  if (is.numeric(color)   && length(color) == 1)             return("uniform_string")
  if ((is.character(color) || is.factor(color)) && length(color) == n) return("categorical")
  if (is.numeric(color) && length(color) == n)               return("numeric")
  stop(
    "color must be NULL, a single color string, a character/factor vector of length n, ",
    "or a numeric vector of length n.",
    call. = FALSE
  )
}

# ---- Main function -----------------------------------------------------------

#' Pavlab-style scatter plot (ggplot2).
#'
#' Opinionated lab defaults: black filled circles, no background, no gridlines,
#' top + right axes off (theme_classic). N-adaptive rendering. Returns a ggplot2
#' object that can be extended with `+` layers.
#'
#' @param x,y Numeric vectors of equal length. Non-finite pairs are silently
#'   dropped.
#' @param color NULL (all black), a single color string (uniform), a
#'   character / factor vector of length n (categorical — lab accent palette +
#'   legend), or a numeric vector of length n (viridis colorbar). Ignored with
#'   a warning when N > 10 000 (density mode).
#' @param color_label Colorbar title for numeric `color`. Default NULL (empty).
#' @param label_size "small", "med", "large", or a numeric point size for axis
#'   labels (tick labels scale proportionally). Default "med" (14 pt / 11 pt).
#' @param xlabel,ylabel Axis labels. Log-scale suffix (“(log₂)” etc.) appended
#'   automatically when a transform is applied.
#' @param title Plot title. Left-aligned, normal weight.
#' @param log_x,log_y FALSE | TRUE | "log2" | "log10". TRUE auto-selects:
#'   log10 when max > 1 000 (counts), log2 otherwise (expression). pseudocount
#'   added before transform so zeros survive. Warns when data spans > 100x and
#'   no transform is requested. ln is never used.
#' @param pseudocount Added to data before log transform. Default 1.
#' @param origin_zero If TRUE (default) and all transformed data >= 0, expands
#'   the axis lower bound to include 0. Explicit xlim / ylim take precedence.
#' @param xlim,ylim Numeric length-2 vectors for explicit axis limits. Take
#'   precedence over origin_zero.
#' @param point_size Override auto marker size (ggplot2 `size=` in mm).
#'   Default: 2.0 for N <= 500, 1.5 for N 501-10 000.
#' @param filename Save path. Format inferred from extension via ggsave.
#' @param figsize Numeric length-2 vector c(width, height) in inches.
#'   Default c(5.5, 5.0).
#' @return ggplot2 object (invisibly).
#' @export
pavlab_scatter <- function(
  x, y,
  color       = NULL,
  color_label = NULL,
  label_size  = "med",
  xlabel      = NULL,
  ylabel      = NULL,
  title       = NULL,
  log_x       = FALSE,
  log_y       = FALSE,
  pseudocount = 1.0,
  origin_zero = TRUE,
  xlim        = NULL,
  ylim        = NULL,
  point_size  = NULL,
  filename    = NULL,
  figsize     = NULL
) {
  if (!requireNamespace("ggplot2", quietly = TRUE))
    stop("pavlab_scatter: ggplot2 is required. Install with install.packages('ggplot2').",
         call. = FALSE)

  x <- as.numeric(x)
  y <- as.numeric(y)
  if (length(x) != length(y))
    stop(sprintf("x and y must have equal length (%d vs %d)", length(x), length(y)),
         call. = FALSE)

  # Align per-point color, then drop non-finite pairs.
  has_vec_color <- !is.null(color) && length(color) == length(x)
  mask <- is.finite(x) & is.finite(y)
  if (has_vec_color) color <- color[mask]
  x <- x[mask]
  y <- y[mask]

  n <- length(x)
  if (n == 0)
    warning("pavlab_scatter: no finite data points to plot.", call. = FALSE)

  # Suggest log scale when not transforming
  if (identical(log_x, FALSE)) .maybe_warn_log(x, "x")
  if (identical(log_y, FALSE)) .maybe_warn_log(y, "y")

  # Apply log transforms
  x_kind <- .pick_log_kind(log_x, x, "x")
  if (!is.null(x_kind)) {
    neg <- sum(is.finite(x) & (x + pseudocount) <= 0)
    if (neg > 0)
      warning(sprintf(
        "pavlab_scatter: %d x value(s) will produce -Inf after log transform (pseudocount=%g insufficient).",
        neg, pseudocount
      ), call. = FALSE)
    x <- .apply_log(x, x_kind, pseudocount)
  }

  y_kind <- .pick_log_kind(log_y, y, "y")
  if (!is.null(y_kind)) {
    neg <- sum(is.finite(y) & (y + pseudocount) <= 0)
    if (neg > 0)
      warning(sprintf(
        "pavlab_scatter: %d y value(s) will produce -Inf after log transform (pseudocount=%g insufficient).",
        neg, pseudocount
      ), call. = FALSE)
    y <- .apply_log(y, y_kind, pseudocount)
  }

  # Density mode
  density_mode <- n > .N_DENSITY
  if (density_mode && !is.null(color)) {
    warning("pavlab_scatter: N > 10 000; using hexbin density mode — 'color' ignored.",
            call. = FALSE)
    color <- NULL
  }

  # Font sizes
  fs       <- .font_sizes(label_size)
  label_pt <- fs[1]
  tick_pt  <- fs[2]

  # Axis labels with log suffix
  xl <- .axis_label(xlabel, x_kind)
  yl <- .axis_label(ylabel, y_kind)

  # Point size + alpha
  ps <- if (!is.null(point_size)) as.numeric(point_size) else
          if (n <= .N_ALPHA) 2.0 else 1.5

  alpha_val <- if (n <= .N_ALPHA) {
    1.0
  } else {
    frac <- min(1.0, (n - .N_ALPHA) / max(1L, .N_DENSITY - .N_ALPHA))
    max(0.15, 0.6 - frac * 0.45)
  }

  # ---- Data frame -----------------------------------------------------------
  df <- data.frame(x = x, y = y)

  # ---- Build ggplot ---------------------------------------------------------

  if (density_mode) {
    if (requireNamespace("hexbin", quietly = TRUE)) {
      p <- ggplot2::ggplot(df, ggplot2::aes(x = x, y = y)) +
        ggplot2::geom_hex(bins = 40) +
        ggplot2::scale_fill_gradientn(colors = black_body_palette(256), name = "count")
    } else {
      message("pavlab_scatter: hexbin not installed; using rectangular bin2d. Install hexbin for hexagonal bins.")
      p <- ggplot2::ggplot(df, ggplot2::aes(x = x, y = y)) +
        ggplot2::geom_bin2d(bins = 40) +
        ggplot2::scale_fill_gradientn(colors = black_body_palette(256), name = "count")
    }

  } else {
    cmode <- .color_mode(color, n)

    if (cmode == "uniform_black") {
      p <- ggplot2::ggplot(df, ggplot2::aes(x = x, y = y)) +
        ggplot2::geom_point(
          color = "#000000", alpha = alpha_val, size = ps, shape = 16, stroke = 0
        )

    } else if (cmode == "uniform_string") {
      p <- ggplot2::ggplot(df, ggplot2::aes(x = x, y = y)) +
        ggplot2::geom_point(
          color = as.character(color), alpha = alpha_val, size = ps,
          shape = 16, stroke = 0
        )

    } else if (cmode == "categorical") {
      cats     <- unique(as.character(color))
      cat_idx  <- ((seq_along(cats) - 1L) %% length(.ACCENT_COLORS)) + 1L
      cat_cols <- setNames(.ACCENT_COLORS[cat_idx], cats)
      df$color_col <- as.character(color)
      p <- ggplot2::ggplot(df, ggplot2::aes(x = x, y = y, color = color_col)) +
        ggplot2::geom_point(alpha = alpha_val, size = ps, shape = 16, stroke = 0) +
        ggplot2::scale_color_manual(values = cat_cols, name = NULL) +
        ggplot2::guides(color = ggplot2::guide_legend(
          override.aes = list(size = 3, alpha = 1)
        ))

    } else {
      # numeric → viridis colorbar
      df$color_col <- as.numeric(color)
      clabel <- if (!is.null(color_label)) color_label else ""
      p <- ggplot2::ggplot(df, ggplot2::aes(x = x, y = y, color = color_col)) +
        ggplot2::geom_point(alpha = alpha_val, size = ps, shape = 16, stroke = 0) +
        ggplot2::scale_color_viridis_c(name = clabel)
    }
  }

  # ---- Theme: no background, no gridlines, top+right off -------------------
  p <- p +
    ggplot2::theme_classic(base_size = tick_pt) +
    ggplot2::theme(
      axis.title        = ggplot2::element_text(size = label_pt, color = "#1f2937"),
      axis.text         = ggplot2::element_text(size = tick_pt,  color = "#6b7280"),
      axis.ticks        = ggplot2::element_line(color = "#6b7280"),
      plot.title        = ggplot2::element_text(size = label_pt, color = "#1f2937",
                                                hjust = 0, face = "plain"),
      legend.text       = ggplot2::element_text(size = tick_pt),
      legend.title      = ggplot2::element_text(size = tick_pt),
      legend.background = ggplot2::element_blank(),
      legend.key        = ggplot2::element_blank(),
      plot.margin       = ggplot2::margin(3, 3, 3, 3, "mm")
    )

  # ---- Labels ---------------------------------------------------------------
  if (nchar(xl) > 0) p <- p + ggplot2::xlab(xl)
  if (nchar(yl) > 0) p <- p + ggplot2::ylab(yl)
  if (!is.null(title)) p <- p + ggplot2::ggtitle(title)

  # ---- Axis limits ----------------------------------------------------------
  # expand_limits trains the scale to include 0; coord_cartesian clips the view.
  if (origin_zero) {
    fx <- x[is.finite(x)]
    if (length(fx) > 0 && min(fx) >= 0) p <- p + ggplot2::expand_limits(x = 0)
    fy <- y[is.finite(y)]
    if (length(fy) > 0 && min(fy) >= 0) p <- p + ggplot2::expand_limits(y = 0)
  }
  if (!is.null(xlim) || !is.null(ylim)) {
    p <- p + ggplot2::coord_cartesian(xlim = xlim, ylim = ylim)
  }

  # ---- Save -----------------------------------------------------------------
  if (!is.null(filename)) {
    w <- if (!is.null(figsize)) figsize[1] else 5.5
    h <- if (!is.null(figsize)) figsize[2] else 5.0
    ggplot2::ggsave(filename, plot = p, width = w, height = h, dpi = 150)
  }

  invisible(p)
}
